File: src/analytics/forecast.py
"""Money-flow forecasting (config/forecast.json).

Models daily **income** and **expense** separately as harmonic regressions —
[intercept, linear trend, weekly + monthly Fourier terms] — fitted by
exponentially time-decayed weighted least squares (statsmodels WLS) so recent
behaviour dominates. Forecasts the future cumulative balance from the current net
worth, with uncertainty bands that grow with the horizon (driven by both income
and expense volatility).

Only the fit (`train_model`) needs statsmodels; prediction works from the stored
coefficients/covariance with numpy, so the model persists as plain JSON.
"""

import json
from datetime import date, datetime
from pathlib import Path

import numpy as np
import pandas as pd

MODEL_PATH = Path("config/forecast.json")

# Signed contribution of each transaction type to net worth (mirrors money_flow).
_SIGN = {"Income": 1, "Transfer-In": 1, "Income Balance": 1,
         "Expense": -1, "Transfer-Out": -1, "Expense Balance": -1}

_WEEKLY, _MONTHLY, _ANNUAL = 7.0, 30.44, 365.25

DEFAULT_CFG = {
    "half_life_days": 120,
    "harmonics": {"weekly": 2, "monthly": 2, "annual": 0},
}

_Z = {"50": 0.6745, "90": 1.6449}  # normal quantiles for the two bands


# ── Design matrix ─────────────────────────────────────────────────────────────

def _design_matrix(t: np.ndarray, cfg: dict) -> np.ndarray:
    """Columns [1, trend, weekly/monthly/annual Fourier] for day indices ``t``
    (days since the fit start). Column order is fixed by ``cfg`` so train and
    predict agree."""
    t = np.asarray(t, dtype=float)
    cols = [np.ones_like(t), t / 365.0]
    h = cfg.get("harmonics", DEFAULT_CFG["harmonics"])
    for period, key in ((_WEEKLY, "weekly"), (_MONTHLY, "monthly"), (_ANNUAL, "annual")):
        for k in range(1, int(h.get(key, 0)) + 1):
            ang = 2.0 * np.pi * k * t / period
            cols.append(np.sin(ang))
            cols.append(np.cos(ang))
    return np.column_stack(cols)


# ── Data helpers ──────────────────────────────────────────────────────────────

def _daily_series(df: pd.DataFrame, kind: str, start, end) -> np.ndarray:
    """Daily-summed amounts for ``kind`` ("Income"/"Expense"), 0-filled over the
    inclusive day range [start, end]."""
    idx = pd.date_range(start, end, freq="D")
    sub = df[df["Income/Expense"] == kind]
    if sub.empty:
        return np.zeros(len(idx))
    daily = sub.groupby(sub["Period"].dt.normalize())["Amount"].sum()
    return daily.reindex(idx, fill_value=0.0).to_numpy(dtype=float)


def _net_worth(df: pd.DataFrame) -> float:
    s = df["Income/Expense"].map(_SIGN).fillna(0.0) * df["Amount"]
    return float(s.sum())


def _fit_one(X: np.ndarray, y: np.ndarray, w: np.ndarray) -> dict:
    """Weighted least squares for one series → coef / cov / residual variance."""
    import statsmodels.api as sm
    res = sm.WLS(y, X, weights=w).fit()
    return {"coef": res.params.tolist(),
            "cov": np.asarray(res.cov_params()).tolist(),
            "sigma2": float(res.scale)}


# ── Train / persist ───────────────────────────────────────────────────────────

def train_model(df: pd.DataFrame, cfg: dict | None = None,
                path: str | Path = MODEL_PATH) -> dict:
    """Fit income & expense models with exponential recency weights and persist."""
    cfg = {**DEFAULT_CFG, **(cfg or {})}
    days = df["Period"].dt.normalize()
    start, end = days.min(), days.max()
    t = np.arange((end - start).days + 1, dtype=float)
    X = _design_matrix(t, cfg)

    # Most-recent day has weight 1; older days decay by the half-life.
    age = t[-1] - t
    w = 0.5 ** (age / float(cfg["half_life_days"]))

    inc = _fit_one(X, _daily_series(df, "Income", start, end), w)
    exp = _fit_one(X, _daily_series(df, "Expense", start, end), w)

    model = {
        "trained_at": datetime.now().isoformat(timespec="seconds"),
        "fit_start": start.date().isoformat(),
        "cfg": cfg,
        "income": inc,
        "expense": exp,
    }
    save_model(model, path)
    return model


def save_model(model: dict, path: str | Path = MODEL_PATH) -> None:
    path = Path(path)
    path.parent.mkdir(parents=True, exist_ok=True)
    with open(path, "w", encoding="utf-8") as f:
        json.dump(model, f, indent=2, ensure_ascii=False)


def load_model(path: str | Path = MODEL_PATH) -> dict | None:
    path = Path(path)
    if not path.exists():
        return None
    with open(path, "r", encoding="utf-8") as f:
        return json.load(f)


# ── Forecast ──────────────────────────────────────────────────────────────────

def forecast(df: pd.DataFrame, model: dict, horizon_days: int) -> dict:
    """Future cumulative-balance fan, anchored at the current net worth.

    Returns ``{dates, median, lo50, hi50, lo90, hi90, anchor_date, anchor_value}``
    with the anchor included as day 0 so the fan starts at today's balance.
    """
    cfg = model.get("cfg", DEFAULT_CFG)
    fit_start = pd.Timestamp(model["fit_start"])
    anchor_date = df["Period"].dt.normalize().max()
    anchor_value = _net_worth(df)

    inc, exp = model["income"], model["expense"]
    b_inc, b_exp = np.array(inc["coef"]), np.array(exp["coef"])
    cov_inc, cov_exp = np.array(inc["cov"]), np.array(exp["cov"])
    s2_inc, s2_exp = float(inc["sigma2"]), float(exp["sigma2"])

    anchor_idx = (anchor_date - fit_start).days
    future_t = np.arange(anchor_idx + 1, anchor_idx + int(horizon_days) + 1, dtype=float)
    Xf = _design_matrix(future_t, cfg)

    # Daily means (income & expense are non-negative → clamp), net = inc − exp.
    mu_inc = np.maximum(Xf @ b_inc, 0.0)
    mu_exp = np.maximum(Xf @ b_exp, 0.0)
    net = mu_inc - mu_exp

    dates = [anchor_date]
    median = [anchor_value]
    lo50, hi50, lo90, hi90 = [anchor_value], [anchor_value], [anchor_value], [anchor_value]

    cum = anchor_value
    resid_var = 0.0
    Sx = np.zeros(Xf.shape[1])
    for i in range(len(future_t)):
        cum += net[i]
        resid_var += s2_inc + s2_exp
        Sx += Xf[i]
        param_var = float(Sx @ cov_inc @ Sx + Sx @ cov_exp @ Sx)
        sd = float(np.sqrt(max(0.0, resid_var + param_var)))
        dates.append(anchor_date + pd.Timedelta(days=i + 1))
        median.append(cum)
        lo50.append(cum - _Z["50"] * sd)
        hi50.append(cum + _Z["50"] * sd)
        lo90.append(cum - _Z["90"] * sd)
        hi90.append(cum + _Z["90"] * sd)

    return {
        "dates": dates, "median": median,
        "lo50": lo50, "hi50": hi50, "lo90": lo90, "hi90": hi90,
        "anchor_date": anchor_date, "anchor_value": anchor_value,
        "trained_at": model.get("trained_at"),
    }

File: web/src/lib/analytics/forecast.ts
// Money-flow forecasting — a browser port of src/analytics/forecast.py.
//
// Daily income and expense are each modelled as a harmonic regression
// [intercept, linear trend, weekly + monthly Fourier terms] fitted by
// exponentially time-decayed weighted least squares so recent behaviour
// dominates. The future cumulative balance is projected from the current net
// worth, with 50% / 90% bands that widen with the horizon (from both income and
// expense volatility).
//
// The Python fit used statsmodels WLS; here WLS is solved directly from the
// normal equations (a small Gauss–Jordan inverse), so there is no heavy
// dependency and the whole thing runs on-device. The model is refit from the
// ledger on demand rather than persisted — it is cheap for a personal ledger.
import { signedAmount } from '../balances'
import type { Txn } from '../../db'

const WEEKLY = 7.0
const MONTHLY = 30.44
const ANNUAL = 365.25

export interface Harmonics {
  weekly: number
  monthly: number
  annual: number
}
export interface ForecastCfg {
  halfLifeDays: number
  harmonics: Harmonics
}
export const DEFAULT_CFG: ForecastCfg = {
  halfLifeDays: 120,
  harmonics: { weekly: 2, monthly: 2, annual: 0 },
}

// Normal quantiles for the two uncertainty bands (mirrors forecast.py `_Z`).
const Z50 = 0.6745
const Z90 = 1.6449

// ── Small dense linear algebra (fixed, ~10×10 at most) ───────────────────────

// Invert a square matrix by Gauss–Jordan elimination with partial pivoting.
function invert(a: number[][]): number[][] {
  const n = a.length
  // Work on [A | I].
  const m = a.map((row, i) => [...row, ...Array.from({ length: n }, (_, j) => (i === j ? 1 : 0))])
  for (let col = 0; col < n; col++) {
    // Pivot: largest magnitude in this column at/below the diagonal.
    let piv = col
    for (let r = col + 1; r < n; r++) if (Math.abs(m[r][col]) > Math.abs(m[piv][col])) piv = r
    if (Math.abs(m[piv][col]) < 1e-12) throw new Error('singular matrix')
    ;[m[col], m[piv]] = [m[piv], m[col]]
    const d = m[col][col]
    for (let j = 0; j < 2 * n; j++) m[col][j] /= d
    for (let r = 0; r < n; r++) {
      if (r === col) continue
      const f = m[r][col]
      if (f === 0) continue
      for (let j = 0; j < 2 * n; j++) m[r][j] -= f * m[col][j]
    }
  }
  return m.map((row) => row.slice(n))
}

const matVec = (a: number[][], v: number[]): number[] => a.map((row) => row.reduce((s, x, j) => s + x * v[j], 0))
const dot = (a: number[], b: number[]): number => a.reduce((s, x, i) => s + x * b[i], 0)

// ── Design matrix ────────────────────────────────────────────────────────────

// Columns [1, trend, weekly/monthly/annual Fourier] for day indices `t` (days
// since the fit start). Column order is fixed by `cfg` so fit and predict agree.
function designMatrix(t: number[], cfg: ForecastCfg): number[][] {
  const h = cfg.harmonics
  const bands: Array<[number, number]> = [
    [WEEKLY, h.weekly],
    [MONTHLY, h.monthly],
    [ANNUAL, h.annual],
  ]
  return t.map((ti) => {
    const cols = [1, ti / 365.0]
    for (const [period, count] of bands) {
      for (let k = 1; k <= count; k++) {
        const ang = (2.0 * Math.PI * k * ti) / period
        cols.push(Math.sin(ang), Math.cos(ang))
      }
    }
    return cols
  })
}

// ── Weighted least squares ───────────────────────────────────────────────────

interface Fit {
  coef: number[]
  cov: number[][] // scaled covariance of the coefficients
  sigma2: number // residual variance (WLS scale = weighted RSS / (n − p))
}

// Solve minₐ Σ wᵢ (yᵢ − Xᵢ·a)² from the normal equations XᵀWX a = XᵀWy.
// cov = σ²(XᵀWX)⁻¹, σ² = Σ wᵢ rᵢ² / (n − p) — matching statsmodels WLS.
function wls(X: number[][], y: number[], w: number[]): Fit {
  const n = X.length
  const p = X[0].length
  const xtwx: number[][] = Array.from({ length: p }, () => new Array(p).fill(0))
  const xtwy: number[] = new Array(p).fill(0)
  for (let i = 0; i < n; i++) {
    const wi = w[i]
    const row = X[i]
    for (let a = 0; a < p; a++) {
      const wr = wi * row[a]
      xtwy[a] += wr * y[i]
      for (let b = 0; b < p; b++) xtwx[a][b] += wr * row[b]
    }
  }
  const inv = invert(xtwx)
  const coef = matVec(inv, xtwy)
  let wrss = 0
  for (let i = 0; i < n; i++) {
    const r = y[i] - dot(X[i], coef)
    wrss += w[i] * r * r
  }
  const sigma2 = n > p ? wrss / (n - p) : 0
  const cov = inv.map((r) => r.map((v) => v * sigma2))
  return { coef, cov, sigma2 }
}

// ── Data helpers ─────────────────────────────────────────────────────────────

const MS_PER_DAY = 86_400_000
const dayNum = (iso: string): number => Math.round(new Date(iso.slice(0, 10) + 'T00:00:00Z').getTime() / MS_PER_DAY)

// Daily-summed amounts for one type, 0-filled over the inclusive [start,end] day
// range (indices are days since `start`).
function dailySeries(txns: Txn[], type: 'Income' | 'Expense', start: number, end: number): number[] {
  const out = new Array(end - start + 1).fill(0)
  for (const t of txns) {
    if (t.type !== type) continue
    const d = dayNum(t.period) - start
    if (d >= 0 && d < out.length) out[d] += t.amount
  }
  return out
}

// ── Public forecast ──────────────────────────────────────────────────────────

export interface Forecast {
  dates: string[] // ISO days, anchor first
  median: number[]
  lo50: number[]
  hi50: number[]
  lo90: number[]
  hi90: number[]
  anchorDate: string
  anchorValue: number
}

const isoDay = (n: number): string => new Date(n * MS_PER_DAY).toISOString().slice(0, 10)

// Fit income & expense from the whole ledger and project the cumulative balance
// `horizonDays` into the future. The fan is anchored at today's net worth and
// begins the next day. Returns null when there is too little history to fit.
export function forecast(txns: Txn[], horizonDays = 30, cfg: ForecastCfg = DEFAULT_CFG): Forecast | null {
  if (txns.length === 0) return null
  const days = txns.map((t) => dayNum(t.period))
  const start = Math.min(...days)
  const end = Math.max(...days)
  const span = end - start
  // Need at least a couple of weeks and more days than parameters to fit.
  const p = designMatrix([0], cfg)[0].length
  if (span < 14 || span + 1 <= p) return null

  const t = Array.from({ length: span + 1 }, (_, i) => i)
  const X = designMatrix(t, cfg)
  // Most-recent day weight 1; older days decay by the half-life.
  const w = t.map((ti) => 0.5 ** ((span - ti) / cfg.halfLifeDays))

  const inc = wls(X, dailySeries(txns, 'Income', start, end), w)
  const exp = wls(X, dailySeries(txns, 'Expense', start, end), w)

  const anchorValue = netWorth(txns)
  const anchorIdx = span // today = last trained day
  const anchorDate = isoDay(start + anchorIdx)

  const futureT = Array.from({ length: horizonDays }, (_, i) => anchorIdx + 1 + i)
  const Xf = designMatrix(futureT, cfg)
  const muInc = matVec(Xf, inc.coef).map((v) => Math.max(v, 0))
  const muExp = matVec(Xf, exp.coef).map((v) => Math.max(v, 0))

  const dates = [anchorDate]
  const median = [anchorValue]
  const lo50 = [anchorValue]
  const hi50 = [anchorValue]
  const lo90 = [anchorValue]
  const hi90 = [anchorValue]

  let cum = anchorValue
  let residVar = 0
  const Sx = new Array(p).fill(0)
  for (let i = 0; i < horizonDays; i++) {
    cum += muInc[i] - muExp[i]
    residVar += inc.sigma2 + exp.sigma2
    for (let j = 0; j < p; j++) Sx[j] += Xf[i][j]
    const paramVar = quad(Sx, inc.cov) + quad(Sx, exp.cov)
    const sd = Math.sqrt(Math.max(0, residVar + paramVar))
    dates.push(isoDay(start + anchorIdx + 1 + i))
    median.push(cum)
    lo50.push(cum - Z50 * sd)
    hi50.push(cum + Z50 * sd)
    lo90.push(cum - Z90 * sd)
    hi90.push(cum + Z90 * sd)
  }

  return { dates, median, lo50, hi50, lo90, hi90, anchorDate, anchorValue }
}

// Net worth = Σ signed amount (transfers cancel across legs).
function netWorth(txns: Txn[]): number {
  return txns.reduce((s, t) => s + signedAmount(t.type, t.amount), 0)
}

// vᵀ M v — the parameter-variance quadratic form.
function quad(v: number[], m: number[][]): number {
  return dot(v, matVec(m, v))
}

// Exposed for tests.
export const _internal = { designMatrix, wls, invert }
